src/test/unit/algorithms/TemporalMemoryTest.cpp

```diff
@@ -0,0 +1,943 @@
+    cell.idx = -1;
+    EXPECT_THROW(tm.columnForCell(cell), std::exception);
+  }
+
+  void TemporalMemoryTest::testCellsForColumn1D()
+  {
+    TemporalMemory tm;
+    tm.initialize(vector<UInt>{2048}, 5);
+
+    vector<Cell> expectedCells = { Cell(5), Cell(6), Cell(7), Cell(8), Cell(9) };
+    vector<Cell> cellsForColumn = tm.cellsForColumn(1);
+    NTA_CHECK(check_vector_eq(cellsForColumn, expectedCells));
+  }
+
+  void TemporalMemoryTest::testCellsForColumn2D()
+  {
+    TemporalMemory tm;
+    tm.initialize(vector<UInt>{64, 64}, 4);
+
+    vector<Cell> expectedCells = { Cell(256), Cell(257), Cell(258), Cell(259) };
+    vector<Cell> cellsForColumn = tm.cellsForColumn(64);
+    NTA_CHECK(check_vector_eq(cellsForColumn, expectedCells));
+  }
+
+  void TemporalMemoryTest::testCellsForColumnInvalidColumn()
+  {
+    TemporalMemory tm;
+    tm.initialize(vector<UInt>{64, 64}, 4);
+
+    EXPECT_NO_THROW(tm.cellsForColumn(4095));
+    EXPECT_THROW(tm.cellsForColumn(4096), std::exception);
+    EXPECT_THROW(tm.cellsForColumn(-1), std::exception);
+  }
+
+  void TemporalMemoryTest::testNumberOfColumns()
+  {
+    TemporalMemory tm;
+    tm.initialize(vector<UInt>{64, 64}, 32);
+
+    int numOfColumns = tm.numberOfColumns();
+    ASSERT_EQ(numOfColumns, 64 * 64);
+  }
+
+  void TemporalMemoryTest::testNumberOfCells()
+  {
+    TemporalMemory tm;
+    tm.initialize(vector<UInt>{64, 64}, 32);
+
+    Int numberOfCells = tm.numberOfCells();
+    ASSERT_EQ(numberOfCells, 64 * 64 * 32);
+  }
+
+  void TemporalMemoryTest::testMapCellsToColumns()
+  {
+    TemporalMemory tm;
+    tm.initialize(vector<UInt>{100}, 4);
+
+    set<Cell> cells = { Cell(0), Cell(1), Cell(2), Cell(5), Cell(399) };
+    map<Int, set<Cell>> columnsForCells = tm.mapCellsToColumns(cells);
+
+    set<Cell> expectedCells = { Cell(0), Cell(1), Cell(2) };
+    NTA_CHECK(check_set_eq(columnsForCells[0], expectedCells));
+    expectedCells = { Cell(5) };
+    NTA_CHECK(check_set_eq(columnsForCells[1], expectedCells));
+    expectedCells = { Cell(399) };
+    NTA_CHECK(check_set_eq(columnsForCells[99], expectedCells));
+  }
+
+  void TemporalMemoryTest::testSaveLoad()
+  {
+    const char* filename = "TemporalMemorySerialization.tmp";
+    TemporalMemory tm1, tm2;
+    vector<UInt> columnDim;
+    UInt numColumns = 12;
+
+    columnDim.push_back(numColumns);
+    tm1.initialize(columnDim);
+
+    ofstream outfile;
+    outfile.open(filename, ios::binary);
+    tm1.save(outfile);
+    outfile.close();
+
+    ifstream infile(filename, ios::binary);
+    tm2.load(infile);
+    infile.close();
+
+    check_spatial_eq(tm1, tm2);
+
+    int ret = ::remove(filename);
+    NTA_CHECK(ret == 0) << "Failed to delete " << filename;
+  }
+
+  void TemporalMemoryTest::testWrite()
+  {
+    const char* filename = "TemporalMemorySerialization.tmp";
+    TemporalMemory tm1, tm2;
+
+    tm1.initialize({ 100 }, 4, 7, 0.37, 0.58, 4, 18, 0.23, 0.08, 91);
+    /*
```

Comment:
Please file a separate issue to finish the rest of this test, and reference it here.